File: sidebars/herdr.swift
// cmux-herdr plugin sidebar (`herdr`).
// Restricted Swift subset (cmux docs/custom-sidebars.md). Root is a view
// expression, not a struct. Bind only live cmux context: workspaces, tabs,
// statuses, agents, progress, git, color. No invented team or fake rows.
// Chrome uses Ghostty/cmux theme tokens so dark/light follow the host.

func hasText(_ value) -> Bool {
  return value != nil && value != ""
}

func hasColor(_ w) -> Bool {
  return hasText(w.color)
}

func hasBranch(_ w) -> Bool {
  return hasText(w.branch)
}

func hasProgress(_ w) -> Bool {
  return w.progress != nil && w.progress.value != nil
}

func hasStatuses(_ w) -> Bool {
  return w.statuses != nil && w.statuses.count > 0
}

func hasAgents(_ w) -> Bool {
  return w.agents != nil && w.agents.count > 0
}

func hasTabs(_ w) -> Bool {
  return w.tabs != nil && w.tabs.count > 0
}

func progressLabel(_ w) -> String {
  if hasProgress(w) && hasText(w.progress.label) {
    return w.progress.label
  }
  if hasProgress(w) {
    let pct = Int(w.progress.value * 100)
    return "\(pct)%"
  }
  return ""
}

func rowTint(_ w) -> String {
  if hasColor(w) { return w.color }
  if w.selected { return "accent" }
  if w.unread > 0 { return "accent" }
  return "secondary"
}

func wsIcon(_ w) -> String {
  let t = w.title.lowercased()
  if t.contains("herdr") { return "rectangle.split.3x3" }
  if w.pinned { return "pin.fill" }
  if w.remote != nil && w.remote.connected == true { return "network" }
  return "folder.fill"
}

func statusTint(_ s) -> String {
  if hasText(s.color) { return s.color }
  return "accent"
}

func statusIcon(_ s) -> String {
  if hasText(s.icon) { return s.icon }
  if hasText(s.key) && s.key.hasPrefix("herdr:") { return "rectangle.split.3x3" }
  return "circle.fill"
}

func statusLabel(_ s) -> String {
  if hasText(s.value) { return s.value }
  if hasText(s.key) { return s.key }
  return ""
}

func agentTint(_ a) -> String {
  if a.status == "working" { return "accent" }
  if a.status == "needs_input" { return "red" }
  if a.status == "ended" { return "tertiary" }
  return "secondary"
}

func agentIcon(_ a) -> String {
  if a.status == "working" { return "hammer.fill" }
  if a.status == "needs_input" { return "exclamationmark.triangle.fill" }
  if a.status == "ended" { return "checkmark.circle" }
  return "pause.circle"
}

func tabFocusId(_ t) -> String {
  if hasText(t.surfaceId) { return t.surfaceId }
  return t.id
}

func herdrStatusChip(_ s) -> some View {
  HStack(spacing: 4) {
    Image(systemName: statusIcon(s))
      .font(.system(size: 8))
      .foregroundColor(statusTint(s))
      .symbolRenderingMode(.hierarchical)
    Text(statusLabel(s))
      .font(.system(size: 10, design: .monospaced))
      .foregroundColor(hasText(s.color) ? s.color : "secondary")
      .lineLimit(1)
  }
  .padding(.horizontal, 5)
  .padding(.vertical, 2)
  .background {
    Capsule().foregroundColor(statusTint(s)).opacity(0.16)
  }
}

func herdrAgentChip(_ a) -> some View {
  HStack(spacing: 4) {
    Image(systemName: agentIcon(a))
      .font(.system(size: 8))
      .foregroundColor(agentTint(a))
      .symbolRenderingMode(.hierarchical)
    Text(a.name)
      .font(.system(size: 10))
      .foregroundColor(agentTint(a))
      .lineLimit(1)
  }
}

func herdrWorkspaceRow(_ w) -> some View {
  Button(action: { cmux("workspace.select", workspace_id: w.id) }) {
    HStack(alignment: .top, spacing: 8) {
      Circle()
        .frame(width: 7, height: 7)
        .foregroundColor(w.selected ? rowTint(w) : "tertiary")
        .padding(.top, 5)
      VStack(alignment: .leading, spacing: 3) {
        HStack(spacing: 6) {
          Image(systemName: wsIcon(w))
            .font(.system(size: 11))
            .foregroundColor(rowTint(w))
            .symbolRenderingMode(.hierarchical)
            .frame(width: 14)
          Text(w.title)
            .font(.system(size: 12))
            .fontWeight(w.selected ? .semibold : .regular)
            .foregroundColor("primary")
            .lineLimit(1)
          Spacer()
          if w.unread > 0 {
            Text("\(w.unread)")
              .font(.system(size: 9, design: .monospaced))
              .fontWeight(.semibold)
              .foregroundColor("accent")
          }
        }
        HStack(spacing: 6) {
          Text("\(w.tabCount) tabs")
            .font(.system(size: 10, design: .monospaced))
            .foregroundColor("secondary")
          if hasBranch(w) {
            Text(w.branch)
              .font(.system(size: 10, design: .monospaced))
              .foregroundColor("tertiary")
              .lineLimit(1)
          }
          if w.dirty == true {
            Text("*")
              .font(.system(size: 10, design: .monospaced))
              .foregroundColor("accent")
          }
          if hasProgress(w) {
            Text(progressLabel(w))
              .font(.system(size: 10, design: .monospaced))
              .foregroundColor("accent")
          }
        }
        if hasProgress(w) && w.progress.value < 1.0 {
          ProgressView(value: w.progress.value, total: 1.0).tint(rowTint(w))
        }
        if hasStatuses(w) {
          HStack(spacing: 4) {
            ForEach(w.statuses.prefix(6)) { s in
              herdrStatusChip(s)
            }
          }
        }
      }
    }
    .padding(.vertical, 5)
    .padding(.horizontal, 6)
    .background {
      RoundedRectangle(cornerRadius: 6)
        .foregroundColor(w.selected ? "accent" : "primary")
        .opacity(w.selected ? 0.14 : 0.04)
    }
  }
}

func herdrTabRow(_ t) -> some View {
  Button(action: { cmux("surface.focus", surface_id: tabFocusId(t)) }) {
    HStack(spacing: 6) {
      Image(systemName: t.focused ? "dot.circle.fill" : "terminal")
        .font(.system(size: 10))
        .foregroundColor(t.focused ? "accent" : "secondary")
        .frame(width: 14)
      Text(t.title)
        .font(.system(size: 11))
        .foregroundColor(t.focused ? "primary" : "secondary")
        .lineLimit(1)
      Spacer()
    }
    .padding(.vertical, 3)
    .padding(.horizontal, 6)
    .background {
      RoundedRectangle(cornerRadius: 4)
        .foregroundColor("primary")
        .opacity(0.04)
    }
  }
}

func selectedWorkspaceDetail(_ w) -> some View {
  VStack(alignment: .leading, spacing: 6) {
    Text("Focused workspace")
      .font(.caption)
      .fontWeight(.semibold)
      .foregroundColor("secondary")
    Text(w.title)
      .font(.system(size: 12))
      .fontWeight(.semibold)
      .foregroundColor("primary")
      .lineLimit(1)
    if hasStatuses(w) {
      Text("Live statuses")
        .font(.system(size: 10))
        .fontWeight(.semibold)
        .foregroundColor("tertiary")
      ForEach(w.statuses.prefix(12)) { s in
        herdrStatusChip(s)
      }
    }
    if hasAgents(w) {
      Text("Live agents")
        .font(.system(size: 10))
        .fontWeight(.semibold)
        .foregroundColor("tertiary")
      ForEach(w.agents.prefix(8)) { a in
        herdrAgentChip(a)
      }
    }
    if hasTabs(w) {
      Text("Surfaces")
        .font(.system(size: 10))
        .fontWeight(.semibold)
        .foregroundColor("tertiary")
      ForEach(w.tabs.prefix(12)) { t in
        herdrTabRow(t)
      }
    }
  }
}

ScrollView {
  VStack(alignment: .leading, spacing: 10) {
    HStack(spacing: 8) {
      Image(systemName: "rectangle.split.3x3")
        .foregroundColor("accent")
        .symbolRenderingMode(.hierarchical)
      Text("Herdr")
        .font(.headline)
        .fontWeight(.semibold)
        .foregroundColor("primary")
      Spacer()
      Text(clock.time)
        .font(.system(size: 10, design: .monospaced))
        .foregroundColor("tertiary")
    }
    if hasText(selectedTitle) {
      Text(selectedTitle)
        .font(.system(size: 11, design: .monospaced))
        .foregroundColor("secondary")
        .lineLimit(1)
    }

    Divider()

    VStack(alignment: .leading, spacing: 6) {
      HStack {
        Text("cmux workspaces")
          .font(.caption)
          .fontWeight(.semibold)
          .foregroundColor("secondary")
        Spacer()
        Text("\(workspaceCount)")
          .font(.system(size: 10, design: .monospaced))
          .foregroundColor("tertiary")
      }
      if workspaces.count == 0 {
        Text("No live cmux workspaces")
          .font(.caption)
          .foregroundColor("tertiary")
      }
      if workspaces.count > 0 {
        Reorderable(workspaces.prefix(40), move: "workspace.reorder") { w in
          herdrWorkspaceRow(w)
        }
      }
    }

    Divider()

    ForEach(workspaces.filter { $0.selected }.prefix(1)) { w in
      selectedWorkspaceDetail(w)
    }

    Divider()

    Text("Status pills update after cmux-herdr sync or watch.")
      .font(.caption)
      .foregroundColor("tertiary")
      .lineLimit(3)
  }
  .padding(12)
}

File: sidebars/herdr.js
// LEGACY CONTRIB FALLBACK — NOT THE PRODUCT INSTALL.
// The product path is: cmux sidebar plugin install|use|update cmux-herdr.
// This file only exists for cmux builds without `cmux sidebar plugin`; copying
// it to ~/.config/cmux/sidebars and running `cmux sidebar select|open herdr`
// replaces the left rail or opens a pane and is unsupported.
// Herdr sidebar for cmux (JS runtime — wins over herdr.swift for live drag).
// Native cmux chrome: Reorderable, context menus, accent/primary/secondary/
// tertiary tokens. Herdr is the product name. This is not an iframe, not a
// bridge panel, and not a CLI cheat-sheet. Click a row to select a workspace
// or focus a surface. Restricted JS scene (cmux docs/custom-sidebars.md).

let selectOverride = null;
const [selectTick, setSelectTick] = signal(0);

function isSelected(w) {
  selectTick();
  if (!w) return false;
  if (selectOverride) {
    if (data.selectedId() === selectOverride) selectOverride = null;
    else return w.id === selectOverride;
  }
  return !!w.selected;
}

function selectWorkspace(id) {
  if (!id) return;
  selectOverride = id;
  setSelectTick(selectTick() + 1);
  cmux("workspace.select", { workspace_id: id });
}

const closedOverride = new Set();
const [closeTick, setCloseTick] = signal(0);

function visibleWorkspaces() {
  closeTick();
  let ws = data.workspaces() ?? [];
  for (const id of Array.from(closedOverride)) {
    if (!ws.some((w) => w.id === id)) closedOverride.delete(id);
  }
  return ws.filter((w) => !closedOverride.has(w.id)).slice(0, 40);
}

function closeWorkspace(id) {
  closedOverride.add(id);
  setCloseTick(closeTick() + 1);
  cmux("workspace.close", { workspace_id: id });
}

function workspaceAction(action, id, extra) {
  const params = { action: action, workspace_id: id };
  if (extra) {
    for (const key of Object.keys(extra)) params[key] = extra[key];
  }
  cmux("workspace.action", params);
}

function tabFocusId(t) {
  if (t && t.surfaceId) return t.surfaceId;
  return t ? t.id : "";
}

function focusSurface(t) {
  const id = tabFocusId(t);
  if (!id) return;
  cmux("surface.focus", { surface_id: id });
}

function hasText(value) {
  return value != null && value !== "";
}

function statusLabel(s) {
  if (s && hasText(s.value)) return s.value;
  if (s && hasText(s.key) && String(s.key).indexOf(":") === -1) return s.key;
  return "";
}

function statusTint(s) {
  if (s && hasText(s.color)) return s.color;
  const label = statusLabel(s).toLowerCase();
  if (label === "working") return "accent";
  if (label === "blocked" || label === "needs_input") return "red";
  if (label === "done" || label === "ended") return "tertiary";
  return "secondary";
}

function statusIcon(s) {
  if (s && hasText(s.icon)) return s.icon;
  const label = statusLabel(s).toLowerCase();
  if (label === "working") return "hammer.fill";
  if (label === "blocked" || label === "needs_input") return "exclamationmark.triangle.fill";
  if (label === "done" || label === "ended") return "checkmark.circle";
  if (label === "idle") return "pause.circle";
  return "circle.fill";
}

function liveStatuses(w) {
  const items = (w && w.statuses) ? w.statuses : [];
  return items.filter((s) => statusLabel(s) !== "").slice(0, 6);
}

function handleMove(id, index) {
  cmux("workspace.reorder", { workspace_id: id, index: index });
}

function workspaceMenu(w) {
  const row = () => w();
  return [
    Button("Rename", () =>
      workspaceAction("rename", row().id)),
    Button(() => (row()?.pinned ? "Unpin" : "Pin"), () =>
      workspaceAction(row()?.pinned ? "unpin" : "pin", row().id)),
    Button(() => (row()?.unread > 0 ? "Mark as Read" : "Mark as Unread"), () =>
      workspaceAction(row()?.unread > 0 ? "mark_read" : "mark_unread", row().id)),
    Divider(),
    Menu("Move", [
      Button("Move Up", () => workspaceAction("move_up", row().id)),
      Button("Move Down", () => workspaceAction("move_down", row().id)),
      Button("Move to Top", () => workspaceAction("move_top", row().id)),
    ]),
    Divider(),
    Button("Close Others", () => workspaceAction("close_others", row().id)).destructive(),
    Button("Close", () => closeWorkspace(row().id)).destructive(),
  ];
}

function statusChip(s) {
  return HStack({ spacing: 4 }, [
    Image(() => statusIcon(s())).font(8).color(() => statusTint(s())),
    Text(() => statusLabel(s())).font(10).color(() => statusTint(s())).lineLimit(1),
  ])
    .paddingHorizontal(5)
    .paddingVertical(2)
    .cornerRadius(7);
}

function workspaceRow(w) {
  const selected = () => isSelected(w());
  return VStack({ spacing: 3 }, [
    HStack({ spacing: 8 }, [
      Circle({ size: 7 }).fill(() => (selected() ? "accent" : "tertiary")),
      Text(() => w()?.title ?? "")
        .font(13)
        .lineLimit(1)
        .truncation("tail")
        .marquee()
        .color(() => (selected() ? "primary" : "secondary")),
      Spacer({ minLength: 0 }),
      Text(() => (w()?.unread > 0 ? String(w().unread) : ""))
        .font(9)
        .weight("semibold")
        .color("accent"),
    ]),
    HStack({ spacing: 6 }, [
      Text(() => String(w()?.tabCount ?? 0)).font(10).color("tertiary"),
      Text(() => (hasText(w()?.branch) ? w().branch : ""))
        .font(10)
        .color("tertiary")
        .lineLimit(1),
      Text(() => (w()?.progress && hasText(w().progress.label) ? w().progress.label : ""))
        .font(10)
        .color("accent")
        .lineLimit(1),
    ]),
    ForEach(
      { items: () => liveStatuses(w()), key: (s) => (s.key || s.value || s.icon || "s") },
      (s) => statusChip(s)
    ),
  ])
    .paddingHorizontal(10)
    .paddingVertical(6)
    .cornerRadius(8)
    .background(() => (selected() ? "#7f7f7f3d" : null))
    .hoverBackground(() => (selected() ? "#7f7f7f3d" : "#7f7f7f24"))
    .frame({ maxWidth: "infinity" })
    .onTap(() => selectWorkspace(w().id))
    .contextMenu(workspaceMenu(w));
}

function tabRow(t) {
  return HStack({ spacing: 6 }, [
    Image(() => (t()?.focused ? "dot.circle.fill" : "terminal"))
      .font(10)
      .color(() => (t()?.focused ? "accent" : "secondary")),
    Text(() => t()?.title ?? "")
      .font(12)
      .lineLimit(1)
      .truncation("tail")
      .color(() => (t()?.focused ? "primary" : "secondary")),
    Spacer({ minLength: 0 }),
  ])
    .paddingHorizontal(10)
    .paddingVertical(5)
    .marginLeading(14)
    .cornerRadius(8)
    .hoverBackground("#7f7f7f24")
    .frame({ maxWidth: "infinity" })
    .onTap(() => focusSurface(t()));
}

function selectedTabs() {
  const ws = visibleWorkspaces();
  const current = ws.find((w) => isSelected(w));
  const tabs = current && current.tabs ? current.tabs : [];
  return tabs.slice(0, 12);
}

sidebar(() =>
  VStack({ spacing: 4 }, [
    HStack({ spacing: 8 }, [
      Text("Herdr").font(13).weight("semibold").color("primary"),
      Spacer(),
      Text(() => String((data.workspaces() ?? []).length)).font(10).color("tertiary"),
    ])
      .paddingHorizontal(10)
      .paddingVertical(6),
    Reorderable(
      {
        items: visibleWorkspaces,
        key: (w) => w.id,
        spacing: 2,
        onMove: handleMove,
      },
      (w) => workspaceRow(w)
    ),
    ForEach({ items: selectedTabs, key: (t) => t.id }, (t) => tabRow(t)),
  ]),
  { surface: "glass" }
);
